django/contrib/gis/db/backends/postgis/pgraster.py

```diff
@@ -45,13 +45,10 @@ def from_pgraster(data):
     pixeltypes = []
     while data:
         # Get pixel type for this band
-        pixeltype, data = chunk(data, 2)
-        pixeltype = unpack('B', pixeltype)[0]
+        pixeltype_with_flags, data = chunk(data, 2)
+        pixeltype_with_flags = unpack('B', pixeltype_with_flags)[0]
 
-        # Subtract nodata byte from band nodata value if it exists
-        has_nodata = pixeltype >= 64
-        if has_nodata:
-            pixeltype -= 64
+        pixeltype = pixeltype_with_flags & BANDTYPE_PIXTYPE_MASK
```

Comment:
I am not sure if this works as expected. Consider the following example:
```python
>>> BANDTYPE_PIXTYPE_MASK = 0x0F
>>> pixeltype_bin = format(7, '04b')
>>> flags = '0100'  # Hasnodata on, offdb off, isnodata off, reserved off
>>> pixeltype_with_flags = pixeltype_bin + flags
>>> pixeltype_with_flags
... '01110100'
>>> int(pixeltype_with_flags, 2) & BANDTYPE_PIXTYPE_MASK
... 4  # Should be 7
```
As I am not very familiar with bitwise operators, I am not sure if my example makes sense. But this is how I understand it.

So maybe something along the following would work better:

```python
pixeltype_with_flags, data = chunk(data, 2)
pixeltype_with_flags = unpack('B', pixeltype_with_flags)[0]
  # Reformat into 8bit binary with leading zeros.
pixeltype_with_flags_bin = format(pixeltype_with_flags, '08b')
# Get 4 first bits, convert to integer pixeltype.
pixeltype = int(pixeltype_with_flags_bin[:4], 2)
# Extract flags from the remaining bits.
is_offdb, has_nodata, is_nodata, reserved_flag = [bool(int(flag)) for flag in pixeltype_with_flags_bin[4:]]
```
This would look like this:
```python
>>> pixeltype_bin = format(7, '04b')  # Create data encoded in 8bit integer for the example.
>>> flags = '0100'
>>> pixeltype_with_flags = int(pixeltype_bin + flags, 2)
>>> pixeltype_with_flags_bin = format(pixeltype_with_flags, '08b')  # Decode the integer into pixeltype and flags.
>>> pixeltype = int(pixeltype_with_flags_bin[:4], 2)  # Get 4 first bits, convert to integer.
>>> is_offdb, has_nodata, is_nodata, reserved_flag = [bool(int(flag)) for flag in pixeltype_with_flags_bin[4:]]   # Get 4 last bits as booleans.
>>> pixeltype
... 7
>>> [is_offdb, has_nodata, is_nodata, reserved_flag]
... [False, True, False, False]  # Offdb, HasNodata, IsNodata, Reserved

```

Reply:
Thanks @yellowcap for reviewing the pull request. I've never before used bitwise operators in python either,  but there are two reasons I did in this case. Firstly, we're manipulating a byte consisting of a number below 16 and flagged with bit flags. Using nibble and flag-masks just makes sense in this domain, mapping it to binary-strings or boolean-arrays and then slicing it seems a bit cumbersome to me. Bitwise ops should perform great as well. Secondly, the specification is expressed in the exact same bit masks and bit flags and so is the postgis implementation (https://github.com/postgis/postgis/blob/svn-trunk/raster/rt_core/rt_serialize.c#L794 ). I tried to keep this fix close to this reference implementation so developers familiar with this domain can read/maintain it.

To get back at your example; you've switched the lower and upper nibble in the byte in pixeltype_with_flags.

>>> BANDTYPE_PIXTYPE_MASK = 0x0F
>>> pixeltype_bin = format(7, '04b')
>>> flags = '0100'  # Hasnodata on, offdb off, isnodata off, reserved off
>>> pixeltype_with_flags = flags + pixeltype_bin  # flags are the upper bits, pixeltype the lower
>>> pixeltype_with_flags
... '01000111'
>>> int(pixeltype_with_flags, 2) & BANDTYPE_PIXTYPE_MASK
... 7  # It now works as expected

Reply:
Thanks @ivorbosloper now it makes sense to me and works fine. I got confused with upper and lower bits wording. Interesting that bits are counted from right to left from that point of view, and that the mask operation works. It would not work the same way with the `BANDTYPE_FLAGS_MASK` as my example shows.